web-common/src/features/models/createDashboardFromModel.ts

```diff
@@ -0,0 +1,213 @@
+import { goto } from "$app/navigation";
+import { notifications } from "@rilldata/web-common/components/notifications";
+import { useDashboardFileNames } from "@rilldata/web-common/features/dashboards/selectors";
+import {
+  getFileAPIPathFromNameAndType,
+  getFilePathFromNameAndType,
+} from "@rilldata/web-common/features/entity-management/entity-mappers";
+import { getName } from "@rilldata/web-common/features/entity-management/name-utils";
+import {
+  createSchemaForTable,
+  ResourceKind,
+  useResource,
+} from "@rilldata/web-common/features/entity-management/resource-selectors";
+import { waitForResource } from "@rilldata/web-common/features/entity-management/resource-status-utils";
+import { EntityType } from "@rilldata/web-common/features/entity-management/types";
+import { generateDashboardYAMLForModel } from "@rilldata/web-common/features/metrics-views/metrics-internal-store";
+import { appScreen } from "@rilldata/web-common/layout/app-store";
+import { overlay } from "@rilldata/web-common/layout/overlay-store";
+import { behaviourEvent } from "@rilldata/web-common/metrics/initMetrics";
+import type { BehaviourEventMedium } from "@rilldata/web-common/metrics/service/BehaviourEventTypes";
+import {
+  MetricsEventScreenName,
+  MetricsEventSpace,
+} from "@rilldata/web-common/metrics/service/MetricsTypes";
+import {
+  RpcStatus,
+  runtimeServicePutFile,
+  V1ReconcileStatus,
+  V1StructType,
+} from "@rilldata/web-common/runtime-client";
+import {
+  createMutation,
+  CreateMutationOptions,
+  MutationFunction,
+  QueryClient,
+  useQueryClient,
+} from "@tanstack/svelte-query";
+import { derived, get } from "svelte/store";
+
+export interface CreateDashboardFromModelRequest {
+  instanceId: string;
+  modelName: string;
+  schema: V1StructType;
+  newDashboardName: string;
+}
+
+export const useCreateDashboardFromModel = <
+  TError = RpcStatus,
+  TContext = unknown
+>(options?: {
+  mutation?: CreateMutationOptions<
+    Awaited<Promise<void>>,
+    TError,
+    { data: CreateDashboardFromModelRequest },
+    TContext
+  >;
+}) => {
+  const { mutation: mutationOptions } = options ?? {};
+  const queryClient = mutationOptions?.queryClient ?? useQueryClient();
+
+  const mutationFn: MutationFunction<
+    Awaited<Promise<void>>,
+    { data: CreateDashboardFromModelRequest }
+  > = async (props) => {
+    const { data } = props ?? {};
+    if (!data.schema)
+      throw new Error("Failed to create dashboard: Model is not ready");
```

Comment:
Instead of adding this safeguard in the action, can we instead add the safeguard at the UI-level by disabling the buttons until the “model has data”?

Reply:
Ya makes sense. I have refactored to use this instead.